Fix computational errors in quaternion implementation and Madgwick IMU filter

deganii opened the pull request:
A few errors were found in the zscilib implementation the Madgwick filter which this pull request addresses.
The python AHRS library was used as a reference for comparison, and the outputs of this implementation were compared with AHRS. The following discrepancies were found:

1) In quaternions.c, the order of operations in zsl_quat_from_ang_vel is reversed. The orientation of the Earth frame relative to the sensor frame must be computed as qw, not wq. One can refer to the AHRS documentation at https://ahrs.readthedocs.io/en/latest/filters/madgwick.html for more details:

<img width="759" alt="Screenshot 2024-09-12 at 6 13 44 PM" src="https://github.com/user-attachments/assets/9d2f1d98-33c3-4d3d-9729-f89e9d12206c">

2) In the Magwick algorithm, we found that the objective function f is computed incorrectly due to an incorrect rotation to the earth frame:

<img width="795" alt="Screenshot_2024-04-01_at_5 01 33 PM" src="https://github.com/user-attachments/assets/920afc04-73ac-42f3-92b2-722396107127">

We fix this and compute f explicitly as per equation 25 of the Madgwick paper. This pull request aligns the implementation with that of AHRS and includes the equation numbers from the original Madgwick paper so that it is easy to troubleshoot.

deganii committed:
Fix computational error in quaternion/Madgwick implementation

A few errors were found in the zscilib implementation the Madgwick filter which this pull request addresses.

The python AHRS library was used as a reference for comparison, and the outputs of this implementation were compared with AHRS. The following discrepancies were found:

1) In quaternions.c, the order of operations in zsl_quat_from_ang_vel is reversed. The orientation of the Earth frame relative to the sensor frame must be computed as qw, not wq. One can refer to the AHRS documentation at https://ahrs.readthedocs.io/en/latest/filters/madgwick.html for more details:

https://github.com/user-attachments/assets/9d2f1d98-33c3-4d3d-9729-f89e9d12206c

2) In the Magwick algorithm, we found that the objective function f is computed incorrectly due to an incorrect rotation to the earth frame:

https://github.com/user-attachments/assets/920afc04-73ac-42f3-92b2-722396107127

We fix this and compute f explicitly as per equation 25 of the Madgwick paper. This pull request aligns the implementation with that of AHRS and includes the equation numbers from the original Madgwick paper so that it is easy to troubleshoot.

Signed-off-by: Ismail Degani <deganii@gmail.com>

microbuilder commented:
Thanks for the fixes. Much appreciated. Can you raise a PR in the Zephyr project that updates the optional west manifest (`submanifests/optional.yaml`), so that the changes can be tested there? 

Let me know if you're not aware of how to do that, but this is the normal process to test sub-modules, replacing the current commit sha with this PR, so something like:

```
    - name: zscilib
      path: modules/lib/zscilib
      revision: pull/62/head
      remote: upstream
      groups:
        - optional
```

And then once this is merged in the zscilib repo, you can update the Zephyr PR to point to the proper commit sha.

microbuilder commented on src/orientation/fusion/madgwick.c:
```diff
@@ -38,31 +38,33 @@ static int zsl_fus_madgwick_imu(struct zsl_vec *g, struct zsl_vec *a,
 	zsl_vec_init(&grad);
 
+
+	// AHRS Convention Variables (https://github.com/Mayitzin/ahrs/)
+	struct zsl_quat q_dot;
+	struct zsl_quat q_gyr;
+
+	q_gyr.r = 0.0;
+	q_gyr.i = g->data[0];
+	q_gyr.j = g->data[1];
+	q_gyr.k = g->data[2];
+
+	// AHRS: qDot = 0.5 * q_prod(q, [0, *gyr]) # (eq. 12)
```
For consistency sake, do you mind using /* ... */ for comments?

deganii replied:
Sure, this is now done

microbuilder commented on src/orientation/fusion/madgwick.c:
```diff
@@ -84,14 +86,21 @@ static int zsl_fus_madgwick_imu(struct zsl_vec *g, struct zsl_vec *a,
-	q->j -= (1.0 / zsl_fus_madg_freq) * (*beta * grad.data[2]);
-	q->k -= (1.0 / zsl_fus_madg_freq) * (*beta * grad.data[3]);
+	/* Update the input quaternion */
+	// AHRS: q += qDot*self.Dt  # (eq. 13)
```
/* ... */

deganii replied:
Fixed

microbuilder commented on src/orientation/fusion/madgwick.c:
```diff
@@ -84,14 +86,21 @@ static int zsl_fus_madgwick_imu(struct zsl_vec *g, struct zsl_vec *a,
 		zsl_vec_from_arr(&grad, jtf.data);
 		zsl_vec_to_unit(&grad);
+
+		// AHRS: qDot -= self.gain*gradient  # (eq. 33)
```
/* ... */

deganii replied:
Fixed

microbuilder commented:
Commits need descriptive comments, and signed-off-by before they can be merged

deganii commented on src/orientation/quaternions.c:
```diff
@@ -509,6 +509,65 @@ int zsl_quat_from_ang_mom(struct zsl_vec *l, struct zsl_quat *qin,
+		// failed to initialize quaternion due to NULL q0
+		rc = -EINVAL;
+		goto err;
+	}
```
All comments here are capitalized and changed to /* ... */

deganii commented:
Thanks for the review. Please see my updates:

- I've modified to a single commit which has all the details of the fixes made
- I've added the signoff requested
- I've raised a top-level commit https://github.com/zephyrproject-rtos/zephyr/pull/78595 which refers to this pull request in the submanifest optional.yaml

microbuilder commented:
Thanks again for the PR. I left a comment on the CI failures on Discord, but looked into it a bit more this morning myself.

There are a number of problems with this commit that you'll need to clean up before it can be merged in, and one that isn't related directly to the PR but that would be good to fix at the same time (in a distinct commit).

I was going to push these changes directly to your fork and branch, but I'm hesitant to do so since you're raising this PR off your`master` branch, and I didn't want to touch that.

## Commit Message

You should comply with Zephyr requirements for commit messages. Here is a rough attempt at fixing that. You can ammend your last commit message with `$ git commit --amend` then for push to your branch to update just the commit message `$ git push -f origin master` (updating for whatever branch names you use):

> The subject line is important to get right in large open source projects like Zephyr due to the quantity of commits involved.

```
orientation: fusion: Fix an error in Madgwick

A few errors were found in the zscilib implementation the Madgwick
filter which this pull request addresses.

The python AHRS library was used as a reference for comparison, and the
outputs of this implementation were compared with AHRS. The following
discrepancies were found:

1) In quaternions.c, the order of operations in zsl_quat_from_ang_vel
   is reversed. The orientation of the Earth frame relative to the
   sensor frame must be computed as qw, not wq. One can refer to the
   AHRS documentation at
   https://ahrs.readthedocs.io/en/latest/filters/madgwick.html for
   more details:

https://github.com/user-attachments/assets/9d2f1d98-33c3-4d3d-9729-f89e9d12206c

2) In the Magwick algorithm, we found that the objective function f is
   computed incorrectly due to an incorrect rotation to the earth frame:

https://github.com/user-attachments/assets/920afc04-73ac-42f3-92b2-722396107127

   We fix this and compute f explicitly as per equation 25 of the
   Madgwick paper. This pull request aligns the implementation with
   that of AHRS and includes the equation numbers from the original
   Madgwick paper so that it is easy to troubleshoot.
```

## madgwick.c Issues

There are a number of issues here. You should be able to catch most of these by running compliance checks, as described in the readme.md in zscilib, and looking at the error messages in the zephyr PR to find the failures: https://github.com/zephyrproject-rtos/zephyr/actions/runs/10911326987/job/30283906252?pr=78595

> You may need to run `$ pip3 install -r requirements-compliance.txt` if you get an error about missing python depdencies

```bash
$ ../../../zephyr/scripts/ci/check_compliance.py \
  -m Gitlint -m Identity -m Nits -m pylint -m checkpatch \
  -c HEAD~2..
```

Here's a diff on some suggestions on how to clean some of these issues up:

```diff
diff --git a/src/orientation/fusion/madgwick.c b/src/orientation/fusion/madgwick.c
index 247ad50..f1bb5f7 100644
--- a/src/orientation/fusion/madgwick.c
+++ b/src/orientation/fusion/madgwick.c
@@ -1,5 +1,5 @@
 /*
-       * Copyright (c) 2021 Marti Riba Pons
+ * Copyright (c) 2021 Marti Riba Pons
  *
  * SPDX-License-Identifier: Apache-2.0
  */
@@ -59,12 +59,13 @@ static int zsl_fus_madgwick_imu(struct zsl_vec *g, struct zsl_vec *a,
                /* Normalize the acceleration vector. */
                zsl_vec_to_unit(a);
 
-               /* Calculate the objective function f by simplifying the sensor field s to
-                * g=[0,0,0,1] and subtracting the acceleration vector. (Equation 25) */
+               /* Calculate the objective function f by simplifying the sensor field s
+                * to g=[0,0,0,1] and subtracting the acceleration vector. (Equation 25)
+                */
                ZSL_MATRIX_DEF(f, 3, 1);
-               f.data[0] = 2.0*(q->i*q->k - q->r *q->j) - a->data[0];
-               f.data[1] = 2.0*(q->r*q->i + q->j *q->k) - a->data[1];
-               f.data[2] = 2.0*(0.5 - q->i*q->i - q->j *q->j) - a->data[2];
+               f.data[0] = 2.0*(q->i*q->k - q->r * q->j) - a->data[0];
+               f.data[1] = 2.0*(q->r*q->i + q->j * q->k) - a->data[1];
+               f.data[2] = 2.0*(0.5 - q->i*q->i - q->j * q->j) - a->data[2];
 
                /* Define and compute the transpose of the Jacobian matrix of the
                 * function f. */
@@ -94,13 +95,13 @@ static int zsl_fus_madgwick_imu(struct zsl_vec *g, struct zsl_vec *a,
                q_dot.k -= *beta * grad.data[3];
        }
 
-       /* Update the input quaternion 
+       /* Update the input quaternion
         * AHRS: q += qDot*self.Dt  # (eq. 13) */
-       zsl_real_t dt = 1.0 / zsl_fus_madg_freq;
-       q -> r += q_dot.r * dt;
-       q -> i += q_dot.i * dt;
-       q -> j += q_dot.j * dt;
-       q -> k += q_dot.k * dt;
+       zsl_real_t dt = (zsl_real_t)1.0 / zsl_fus_madg_freq;
+       q->r += q_dot.r * dt;
+       q->i += q_dot.i * dt;
+       q->j += q_dot.j * dt;
+       q->k += q_dot.k * dt;
 
        /* Normalize the output quaternion. */
        zsl_quat_to_unit_d(q);
```

## Madwick Test Failure

One of the madgwick tests is failing, probably because we need to lower the precision in the value comparison to account for differences between single and double-precision calculations, or small differences across compilers, emulators, etc. Here is a diff of how that can be changed, which should go in it's own commit:

```diff
diff --git a/tests/src/fusion_tests.c b/tests/src/fusion_tests.c
index 55e8a83..72a8bd2 100644
--- a/tests/src/fusion_tests.c
+++ b/tests/src/fusion_tests.c
@@ -906,18 +906,18 @@ ZTEST(zsl_tests, test_fus_kalman)
        /* Run the kalman algorithm. */
        rc = kalm_drv.feed_handler(&a, &m, &g, NULL, &q, kalm_drv.config);
        zassert_true(rc == 0, NULL);
-       zassert_true(val_is_equal(q.r, 0.3186651048490019, 1E-6), NULL);
-       zassert_true(val_is_equal(q.i, 0.6386866767275943, 1E-6), NULL);
-       zassert_true(val_is_equal(q.j, 0.4721142772080947, 1E-6), NULL);
-       zassert_true(val_is_equal(q.k, 0.5173393365852964, 1E-6), NULL);
+       zassert_true(val_is_equal(q.r, 0.3186651048490019, 1E-4), NULL);
+       zassert_true(val_is_equal(q.i, 0.6386866767275943, 1E-4), NULL);
+       zassert_true(val_is_equal(q.j, 0.4721142772080947, 1E-4), NULL);
+       zassert_true(val_is_equal(q.k, 0.5173393365852964, 1E-4), NULL);
 
        /* Run the kalman algorithm with dip angle provided. */
        rc = kalm_drv.feed_handler(&a, &m, &g, &incl, &q, kalm_drv.config);
        zassert_true(rc == 0, NULL);
-       zassert_true(val_is_equal(q.r, 0.7715963603719183, 1E-6), NULL);
-       zassert_true(val_is_equal(q.i, 0.0123858064715810, 1E-6), NULL);
-       zassert_true(val_is_equal(q.j, 0.5043103125990586, 1E-6), NULL);
-       zassert_true(val_is_equal(q.k, 0.3875006542769926, 1E-6), NULL);
+       zassert_true(val_is_equal(q.r, 0.7715963603719183, 1E-4), NULL);
+       zassert_true(val_is_equal(q.i, 0.0123858064715810, 1E-4), NULL);
+       zassert_true(val_is_equal(q.j, 0.5043103125990586, 1E-4), NULL);
+       zassert_true(val_is_equal(q.k, 0.3875006542769926, 1E-4), NULL);
 
        /* Run the kalman algorithm without accelerometer data. An error
         * is expected. */
```

There may be other issues.

I'd also suggest creating a branch for every PR you raise, not basing it off the default brain (`main`, etc.) If every PR has it's own branch, it's much easier to make chages to the code for that PR for maintainers, etc.

microbuilder commented:
@deganii Any update on this? The changes are useful and I'd love to get them merged in.
